werkit/lambda/parallel.py

```diff
@@ -0,0 +1,28 @@
+import boto3
+import json
+import asyncio
+import os
+
+client = boto3.client('lambda')
+
+LAMBDA_WORKER_FUNCTION_NAME = 'LAMBDA_WORKER_FUNCTION_NAME'
+
+if LAMBDA_WORKER_FUNCTION_NAME not in os.environ:
+    raise Error(f'Environment variable {LAMBDA_WORKER_FUNCTION_NAME} must be defined to use s3_upload_handler')
+
+lambda_worker_function_name = os.environ[LAMBDA_WORKER_FUNCTION_NAME]
+
+#TODO: how to handle errors
+async def call_worker_service(y):
+    response = client.invoke(
+        FunctionName=lambda_worker_function_name,
+        Payload=json.dumps(y)
+    )
+    return json.load(response['Payload'])
+
+async def synchronous_map_on_lambda(x):
+    #TODO: deal with object arg
+    coroutines = [call_worker_service(y) for y in x] 
+    completed, pending  = await asyncio.wait(coroutines)
```

Comment:
What does `pending` get here? Is it jobs which have timed out? Probably it's important to do something to handle that case. If nothing else, raise an exception if `pending` is non-empty, so you don't get an array that's randomly missing some of the elements.

Does `completed` appear in the original order, by the way?

Reply:
According to [the docs](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait), this usage (passing coroutine objects directly to wait) is deprecated, and it is better to use `async.gather`

Here is the behaviour I am expecting when calling `async.gather`:

* The call to `async.gather` should return the results, and wait until all coroutines complete.
* Results are returned in list order.
* If one coroutine throws an error, it returns the error as its result. So the calling code is responsible for checking the returned array of objects to see if there are any errors (note that this logic still needs to be added to `parallel_map_on_lambda`)                               
* If one coroutine throws an error, the other coroutines will not be cancelled. Instead they will run until the complete.                                                                     

What I am not sure about:

* If we want calls to sublambdas to timeout. We could do this by wrapping each call to the coroutine in `async.wait_for`. Do we want the orchestrator to control this? We could also simply let the call to the orchestrator lambda timeout and fail.

Reply:
For the sublambdas, I could imagine passing a relatively long timeout parameter (like 120 seconds in our case). That would allow the orchestrator lambda to surface information about which sublambda timed out.